src/util/db.rs

```diff
@@ -359,38 +366,220 @@ pub fn db_mkdirs(path_str: &str) -> Result<(String, String), Error> {
     pub context: C,
     block_linkage: Option<(BlockHeaderHash, BlockHeaderHash)>
 }
 
-impl<'a, C> Deref for IndexDBTx<'a, C> {
+impl<'a, C: Clone> Deref for IndexDBTx<'a, C> {
     type Target = DBTx<'a>;
     fn deref(&self) -> &DBTx<'a> {
-        &self.tx
+        self.tx()
     }
 }
 
-impl<'a, C> DerefMut for IndexDBTx<'a, C> {
+impl<'a, C: Clone> DerefMut for IndexDBTx<'a, C> {
     fn deref_mut(&mut self) -> &mut DBTx<'a> {
-        &mut self.tx
+        self.tx_mut()
+    }
+}
+
+fn tx_busy_handler(run_count: i32) -> bool {
+    let mut sleep_count = 10;
+    if run_count > 0 {
+        sleep_count = sleep_count << ((run_count - 1) as u64);
+        if sleep_count == 0 {
+            sleep_count = u64::max_value();
+        }
+    }
+    sleep_count += thread_rng().gen::<u64>() % sleep_count;
+
+    if sleep_count > 5000 {
+        sleep_count = 5000;
     }
+    debug!("Database is locked; sleeping {}ms and trying again", &sleep_count);
+    sleep_ms(sleep_count);
+    true
 }
 
-impl<'a, C> IndexDBTx<'a, C> {
+/// Begin an immediate-mode transaction, and handle busy errors with exponential backoff.
+/// Handling busy errors when the tx begins is preferable to doing it when the tx commits, since
+/// then we don't have to worry about any extra rollback logic.
+pub fn tx_begin_immediate<'a>(conn: &'a mut Connection) -> Result<DBTx<'a>, Error> {
+    conn.busy_handler(Some(tx_busy_handler)).map_err(Error::SqliteError)?;
+    let tx = Transaction::new(conn, TransactionBehavior::Immediate).map_err(Error::SqliteError)?;
+    Ok(tx)
+}
+
+/// Get the ancestor block hash of a block of a given height, given a descendent block hash.
+pub fn get_ancestor_block_hash(index: &MARF, block_height: u64, tip_block_hash: &BlockHeaderHash) -> Result<Option<BlockHeaderHash>, Error> {
+    assert!(block_height < u32::max_value() as u64);
+    MARF::get_block_at_height(&mut index.reopen_storage_readonly().map_err(Error::IndexError)?, block_height as u32, tip_block_hash).map_err(Error::IndexError)
+}
+
+/// Get the height of an ancestor block, if it is indeed the ancestor.
+pub fn get_ancestor_block_height(index: &MARF, ancestor_block_hash: &BlockHeaderHash, tip_block_hash: &BlockHeaderHash) -> Result<Option<u64>, Error> {
+    match MARF::get_block_height(&mut index.reopen_storage_readonly().map_err(Error::IndexError)?, ancestor_block_hash, tip_block_hash).map_err(Error::IndexError)? {
+        Some(height_u32) => {
+            Ok(Some(height_u32 as u64))
+        }
+        None => {
+            Ok(None)
+        }
+    }
+}
+
+/// Load some index data
+fn load_indexed(conn: &DBConn, marf_value: &MARFValue) -> Result<Option<String>, Error> {
+    let mut stmt = conn.prepare("SELECT value FROM __fork_storage WHERE value_hash = ?1 LIMIT 2").map_err(Error::SqliteError)?;
+    let mut rows = stmt.query(&[&to_hex(&marf_value.to_vec())]).map_err(Error::SqliteError)?;
+    let mut all_values = vec![];
+    while let Some(row_res) = rows.next() {
```

Comment:
If this function is going to panic if `len() > 1`, it shouldn't need to fill a vec. Maybe just fill an option and panic if the option already has a value?

Reply:
Fixed in ef85b0a